### ISSUE TITLE
Detect unexpected errors via logs in envtest and e2e tests

### ISSUE THREAD

**Issue by sbueringer** [kubernetes-sigs/cluster-api#11402]
It would be great if we find a way to surface unexpected conditions via logs and then check in envtest and e2e tests that they never occur.

Of course this requires us to have an easy way to identify such logs (e.g. they could contain a specific keyword).

Recently we implemented something similar for panics. After envtest and e2e tests are completed we check controller-runtime metrics to ensure no panics occured

**Comment by sbueringer** [comment]
/triage accepted

But next step is
> Of course this requires us to have an easy way to identify such logs (e.g. they could contain a specific keyword).

**Comment by faermanj** [comment]
@sbueringer I'm interested in helping with this one. Do we want to catch known patterns (rules based) or anything out of the ordinary (ML-based)?

**Comment by sbueringer** [comment]
Nothing fancy, ideally just looking for a substring. But the problem is this one:
> Of course this requires us to have an easy way to identify such logs (e.g. they could contain a specific keyword).

**Comment by om7057** [comment]
Hi @sbueringer, I'd like to pick this up.

I looked at the existing panic detection (`verifyPanicMetrics` in `internal/test/envtest/environment.go` and `verifyMetrics` in `test/framework/deployment_helpers.go`), and both work off Prometheus counters, so there isn't an existing pattern yet for scanning log lines for a substring. Before starting I'd like to check a couple of things with you:

1. Envtest logs currently go to `GinkgoWriter`/stdout and aren't captured anywhere today, so adding a check there means adding a small log capture step first. E2e already writes controller logs to files through the log watcher, so that side already has something to scan. I'm planning to scope a first PR to envtest only, mirroring how `verifyPanicMetrics` is wired into teardown, and leave e2e log scanning as a follow up since tuning it for log volume and false positives feels like its own piece of work. Does that split sound reasonable?
2. For the substring itself, is there a specific one you had in mind, or would something like a couple of maintainer picked strings to start (with room to add more later) work for a first pass?

I'll wait a bit for your input before starting, and will call out anything I leave out of scope in the PR itself.

**Comment by fabriziopandini** [comment]
/close

The complexity here is in the definition of what an "enexpected error" is, and we don't think it is worth the effort to invest on this now
